pkg/arvo/lib/shoe.hoon

```diff
@@ -0,0 +1,341 @@
+    |~  sole-id=@ta
+    |~(nail *(like command-type))
+  ::
+  ++  tab-list
+    |~  sole-id=@ta
+    ::  (list [@t tank])
+    *(list (option:auto tank))
+  ::
+  ++  on-command
+    |~  [sole-id=@ta command=command-type]
+    *(quip card _^|(..on-init))
+  ::
+  ++  can-connect
+    |~  sole-id=@ta
+    *?
+  ::
+  ++  on-connect
+    |~  sole-id=@ta
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-disconnect
+    |~  sole-id=@ta
+    *(quip card _^|(..on-init))
+  ::
+  ::NOTE  standard gall agent arms below, though they may produce %shoe cards
+  ::
+  ++  on-init
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-save
+    *vase
+  ::
+  ++  on-load
+    |~  vase
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-poke
+    |~  [mark vase]
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-watch
+    |~  path
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-leave
+    |~  path
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-peek
+    |~  path
+    *(unit (unit cage))
+  ::
+  ++  on-agent
+    |~  [wire sign:agent:gall]
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-arvo
+    |~  [wire sign-arvo]
+    *(quip card _^|(..on-init))
+  ::
+  ++  on-fail
+    |~  [term tang]
+    *(quip card _^|(..on-init))
+  --
+::  +default: bare-minimum implementations of shoe arms
+::
+++  default
+  |*  [shoe=* command-type=mold]
+  |_  =bowl:gall
+  ++  command-parser
+    (easy *command-type)
+  ::
+  ++  tab-list
+    ~
+  ::
+  ++  on-command
+    |=  [sole-id=@ta command=command-type]
+    [~ shoe]
+  ::
+  ++  can-connect
+    |=  sole-id=@ta
+    ?|  =(~zod src.bowl)
+        (team:title [our src]:bowl)
+    ==
+  ::
+  ++  on-connect
+    |=  sole-id=@ta
+    [~ shoe]
+  ::
+  ++  on-disconnect
+    |=  sole-id=@ta
+    [~ shoe]
+  --
+::  +agent: creates wrapper core that handles sole events and calls shoe arms
+::
+++  agent
+  |*  command-type=mold
+  |=  =(shoe command-type)
+  =|  state-0
+  =*  state  -
+  ^-  agent:gall
+  =>
+    |%
+    ++  deal
+      |=  cards=(list card)
+      %+  turn  cards
+      |=  =card
+      ^-  card:agent:gall
+      ?.  ?=(%shoe -.card)  card
+      ?-  -.effect.card
+          %sole
+        =-  [%give %fact - %sole-effect !>(effect.effect.card)]
+        %+  turn
+          ?^  sole-ids.card  sole-ids.card
+          ~(tap in ~(key by soles))
+        |=  sole-id=@ta
+        /sole/[sole-id]
+      ==
+    --
+  ::
+  |_  =bowl:gall
+  +*  this  .
+      og    ~(. shoe bowl)
+  ::
+  ++  on-init
+    ^-  (quip card:agent:gall agent:gall)
+    =^  cards  shoe  on-init:og
+    [(deal cards) this]
+  ::
+  ++  on-save   !>([shoe-inner=on-save:og shoe-self=state])
+  ::
+  ++  on-load
+    |=  old-state=vase
+    ^-  (quip card:agent:gall agent:gall)
+    ::  we could be upgrading from a shoe-less app, in which case the vase
+    ::  contains inner application state instead of our +on-save.
+    ::  to distinguish between the two, we check for the presence of our own
+    ::  +on-save faces in the vase.
```

Comment:
RFC, wondering if there's a better way to do this. Having this rely on the faces in `+on-save` feels a bit weird.

Reply:
Agree, case types should generally be *preserved*, not drive logic - dynamically dispatching calls to arms is sometimes the correct move, but even there much of the point of static gall was to do so less. (The double vase also feels weird, though I suppose short of `slop` it'd be hard to get away from.)

How would you feel about an explicit `!>([%shoe-app inner=on-save:og shoe-state=state])` with corresponding 
`?.  ?=([%shoe-app ^])`? Seems as unlikely to collide, and is in a real sense an outer layer state version.

Reply:
> with corresponding `?. ?=([%shoe-app ^])`

How do you unpack (`!<`) to get to be able to do the `?=` though? Or did you mean to do that check as a vase operation also?

Reply:
Sorry yep that's not valid hoon, `?=([%shoe-app ^] q.old-state)`. Having checked the value, we'll know the type lines up(I don't thing we're at risk of getting a fork type), so you don't even have to `sped` just `!<` directly like with the current `worn`.

Reply:
And then like, `=^  old-inner  state  +:!<([%shoe-app vase state-0] old-state)`

Reply:
I think @ohAitch's solution is pretty good.  Although it's cheapest if you know the whole type ahead of time, it's fine to look into a vase's value or even `!<` to more and more specific types, especially when the performance doesn't much matter, as in `+on-load`.  Technically the former is a special case of the latter: `q.vase` is just `!<(* vase)`.

Nested vases are fine if the semantics are actually "this is dynamic data which contains data that is also dynamic even in the context where the original data is static", as is the case here.